torchdata/stateful_dataloader/README.md

```diff
@@ -0,0 +1,215 @@
+(map-style) or the dataset (iterable-style). However if the sampler and/or dataset include `state_dict/load_state_dict`
+methods, then it will call them during its own `state_dict/load_state_dict` calls. Under the hood, StatefulDataLoader
+handles aggregation and distribution of state across multiprocess workers (but not across ranks).
+
+## Installation
+
+`torchdata.stateful_dataloader` is currently available on the `torchdata` Nightly build.
+
+Using pip:
+
+```bash
+pip install --pre torchdata --extra-index-url https://download.pytorch.org/whl/nightly/cpu
+```
+
+Using conda:
+
+```bash
+conda install torchdata -c pytorch-nightly
+```
+
+## Saving and loading state
+
+```py
+from torchdata.stateful_dataloader import StatefulDataLoader
+...
+
+dataloader = StatefulDataLoader(dataset, num_workers=2, ...)
+for i, batch in enumerate(dataloader):
+  ...
+  if i == 10:
+    state_dict = dataloader.state_dict()
+    break
+...
+
+# Training run resumes with the previous checkpoint
+dataloader = StatefulDataLoader(dataset, num_workers=2, ...)
+# Resume state with DataLoader
+dataloader.load_state_dict(state_dict)
+for i, batch in enumerate(dataloader):
+  ...
+
+```
+
+## Saving Custom State with Map-Style Datasets
+
+For efficient resuming, you can resume iteration by defining `state_dict/load_state_dict` methods in your sampler. If
+your dataset has worker-specific state (eg RNG transform state) you can add `state_dict/load_state_dict` methods to your
+dataset.
+
+```py
+from typing import *
+import torch
+import torch.utils.data
+from torchdata.stateful_dataloader import StatefulDataLoader
+
+class MySampler(torch.utils.data.Sampler[int]):
+  def __init__(self, high: int, seed: int, limit: int):
+    self.seed, self.high, self.limit = seed, high, limit
+    self.g = torch.Generator()
+    self.g.manual_seed(self.seed)
+    self.i = 0
+
+  def __iter__(self):
+    while self.i < self.limit:
+      val = int(torch.randint(high=self.high, size=(1,), generator=self.g))
+      self.i += 1
+      yield val
+
+  def load_state_dict(self, state_dict: Dict[str, Any]):
+    self.i = state_dict["i"]
+    self.g.set_state(state_dict["rng"])
+
+  def state_dict(self) -> Dict[str, Any]:
+    return {"i": self.i, "rng": self.g.get_state()}
+
+# Optional: save dataset random transform state
+class NoisyRange(torch.utils.data.Dataset):
+  def __init__(self, high: int, mean: float, std: float):
+    self.high, self.mean, self.std = high, torch.tensor([float(mean)]), float(std)
+
+  def __len__(self):
+    return self.high
+
+  def __getitem__(self, idx: int) -> float:
+    if not (0 <= idx < self.high):
+      raise IndexError()
+    x = torch.normal(self.mean, self.std)
+    noise = x.item()
+    return idx + noise
+
+  def load_state_dict(self, state_dict):
+    torch.set_rng_state(state_dict["rng"])
+
+  def state_dict(self):
+    return {"rng": torch.get_rng_state()}
+
+for num_workers in [0, 2]:
```

Comment:
Maybe add a comment here that it works for both SingleProcess and MultiProcess dataloading setups?